servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -158,4 +164,173 @@ private Single<C> p2c(int size, List<? extends Host<ResolvedAddress, C>> hosts,
     private Random getRandom() {
         return random == null ? ThreadLocalRandom.current() : random;
     }
+
+    private EntrySelector buildAliasTable(List<? extends Host<?, ?>> hosts) {
+        if (hosts.isEmpty()) {
+            return EMPTY_SELECTOR;
+        }
+
+        double[] probs = new double[hosts.size()];
+        boolean allSameProbability = true;
+        double pTotal = 0;
+        for (int i = 0; i < hosts.size(); i++) {
+            final double pi = hosts.get(i).weight();
+            if (pi < 0) {
+                LOGGER.warn("{}: host at address {} has negative weight ({}). Using unweighted selection.",
+                        getTargetResource(), hosts.get(i).address(), pi);
+                return new EqualWeightTable(hosts.size());
+            }
+            probs[i] = pi;
+            pTotal += pi;
+            allSameProbability = allSameProbability && approxEqual(pi, probs[0]);
+        }
+
+        if (allSameProbability) {
+            // no need for a DRV table.
+            return new EqualWeightTable(hosts.size());
+        }
+
+        // make sure our probability is normalized to less than 1% error.
+        if (!approxEqual(pTotal, 1)) {
+            // need to normalize.
+            double invPTotal = 1.0 / pTotal;
+            for (int i = 0; i < probs.length; i++) {
+                probs[i] *= invPTotal;
+            }
+        }
+        return buildAliasTable(probs);
+    }
+
+    private abstract static class EntrySelector {
```

Comment:
I've formed a habit (for better or worse) of using abstract classes for non-public abstractions because virtual dispatch is always going to be as fast or faster than interface dispatch. However, in this case, likely they'll be the same since it's going to be bi-morphic. If you have a strong preference I'm happy to change it.

Reply:
No strong preference, I was just curious. It also doesn't hurt readability so I'm fine sticking with it.